### PR TITLE
[kjobctl] Adds RayJob mode.

### PR DESCRIPTION
**Pull request by mbobrovskyi**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Adds RayJob mode on kjobctl.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: cmd/experimental/kjobctl/pkg/cmd/create/create.go [hunk 1]
```diff
@@ -138,15 +154,31 @@ var createModeSubcommands = map[string]modeSubcommand{
...
+			subcmd.Use += " [--replicas [WORKER_GROUP]=REPLICAS] [--min-replicas [WORKER_GROUP]=MIN_REPLICAS] [--max-replicas [WORKER_GROUP]=MAX_REPLICAS]"
+			subcmd.Short = "Create a rayjob"
+			subcmd.Example = createRayJobExample
+			subcmd.Flags().StringToIntVar(&o.Replicas, replicasFlagName, nil,
+				"Replicas is the number of desired Pods for this worker group.")
+			subcmd.Flags().StringToIntVar(&o.MinReplicas, minReplicasFlagName, nil,
+				"MinReplicas denotes the minimum number of desired Pods for this worker group.")
+			subcmd.Flags().StringToIntVar(&o.MaxReplicas, maxReplicasFlagName, nil,
+				"MaxReplicas denotes the maximum number of desired Pods for this worker group, and the default value is maxInt32.")
+		},
```
Comment (trasc) [inline cmd/experimental/kjobctl/pkg/cmd/create/create.go:180]: We should keep the `cmd` flag for ray as well, a job user ca chose to run a different file from a volume mount provided by the profile.

Without this, just being able run the profile with different replicas is a very limiting UX.

File: cmd/experimental/kjobctl/pkg/builder/ray_job_builder.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+				workerGroupSpec.MinReplicas = ptr.To(int32(minReplicas))
+			}
+			if maxReplicas, ok := b.maxReplicas[workerGroupSpec.GroupName]; ok {
+				workerGroupSpec.MaxReplicas = ptr.To(int32(maxReplicas))
+			}
+
+			workerGroupSpec.Template.Spec = b.buildPodSpec(workerGroupSpec.Template.Spec)
+		}
+	}
+
```
Comment (trasc) [inline cmd/experimental/kjobctl/pkg/builder/ray_job_builder.go:81]: We should propagate the volumes, mounts and env-vars.

File: cmd/experimental/kjobctl/config/samples/application_profile.yaml [hunk 3]
```diff
@@ -45,3 +48,143 @@ template:
...
+            volumes:
+              # You set volumes at the Pod level, then mount them into containers inside that Pod
+              - name: code-sample
+                configMap:
+                  # Provide the name of the ConfigMap you want to mount.
+                  name: ray-job-code-sample
+                  # An array of keys from the ConfigMap to create as files
+                  items:
+                    - key: sample_code.py
+                      path: sample_code.py
```
Comment (trasc) [inline cmd/experimental/kjobctl/config/samples/application_profile.yaml:129]: let's make this mount come from a VolumeBundle.

File: cmd/experimental/kjobctl/config/samples/application_profile.yaml [hunk 4]
```diff
@@ -45,3 +48,143 @@ template:
...
+metadata:
+  name: sample-ray-job-template
+  namespace: default
+template:
+  spec:
+    # submissionMode specifies how RayJob submits the Ray job to the RayCluster.
+    # The default value is "K8sJobMode", meaning RayJob will submit the Ray job via a submitter Kubernetes Job.
+    # The alternative value is "HTTPMode", indicating that KubeRay will submit the Ray job by sending an HTTP request to the RayCluster.
+    # submissionMode: "K8sJobMode"
+    entrypoint: python /home/ray/samples/sample_code.py
```
Comment (trasc) [inline cmd/experimental/kjobctl/config/samples/application_profile.yaml:63]: Use a VolumeBundle env var as a base.

File: cmd/experimental/kjobctl/config/samples/application_profile.yaml [hunk 5]
```diff
@@ -45,3 +48,143 @@ template:
...
+                  resources:
+                    limits:
+                      cpu: "1"
+                    requests:
+                      cpu: "200m"
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: ray-job-code-sample
```
Comment (trasc) [inline cmd/experimental/kjobctl/config/samples/application_profile.yaml:160]: Create a VolumeBundle that will mount this.

**Review by trasc — COMMENTED** [review]
1st pass

File: cmd/experimental/kjobctl/pkg/cmd/create/create.go (hunk 1 above)
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/pkg/cmd/create/create.go:180]: Do you mean entrypoint? 

https://github.com/kubernetes-sigs/kueue/blob/2d971cff71a8d4ac3bcddf7c0f955b24c6ceb704/cmd/experimental/kjobctl/config/samples/application_profile.yaml#L63

File: cmd/experimental/kjobctl/pkg/builder/ray_job_builder.go (hunk 2 above)
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/pkg/builder/ray_job_builder.go:81]: We already have it on buildPodSpec. Or you mean to add to headGroup too? 

https://github.com/kubernetes-sigs/kueue/blob/2d971cff71a8d4ac3bcddf7c0f955b24c6ceb704/cmd/experimental/kjobctl/config/samples/application_profile.yaml#L120-L129

File: cmd/experimental/kjobctl/pkg/cmd/create/create.go (hunk 1 above)
Comment (trasc) [reply cmd/experimental/kjobctl/pkg/cmd/create/create.go:180]: Yes, but for the create command uniformity let's keep naming it cmd

File: cmd/experimental/kjobctl/pkg/builder/ray_job_builder.go (hunk 2 above)
Comment (trasc) [reply cmd/experimental/kjobctl/pkg/builder/ray_job_builder.go:81]: Yes, the way I see it mostly the `head` needs the volume.

File: cmd/experimental/kjobctl/docs/commands/kjobctl_create/_index.md [hunk 6]
```diff
@@ -29,6 +29,16 @@ Create a task
 	--profile my-application-profile  \
 	--pod-running-timeout 30s \
 	--rm
+
+  # Create job 
+  kjobctl create rayjob \ 
+	--profile my-application-profile  \
+	--cmd &#34;sleep 5&#34; \
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/docs/commands/kjobctl_create/_index.md]: Good point. Thanks.

File: cmd/experimental/kjobctl/config/samples/application_profile.yaml [hunk 7]
```diff
@@ -45,3 +48,108 @@ template:
...
+            volumes:
+              # You set volumes at the Pod level, then mount them into containers inside that Pod
+              - name: code-sample
+                configMap:
+                  # Provide the name of the ConfigMap you want to mount.
+                  name: ray-job-code-sample
+                  # An array of keys from the ConfigMap to create as files
+                  items:
+                    - key: ray_sample_code.py
+                      path: ray_sample_code.py
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/config/samples/application_profile.yaml]: Done

File: cmd/experimental/kjobctl/pkg/cmd/create/create.go [hunk 8]
```diff
@@ -60,10 +60,21 @@ const (
 	--profile my-application-profile  \
 	--pod-running-timeout 30s \
 	--rm`
+	createRayJobExample = `  # Create job 
+  kjobctl create rayjob \ 
+	--profile my-application-profile  \
+	--request cpu=500m,ram=4Gi \
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/pkg/cmd/create/create.go]: Done.

File: cmd/experimental/kjobctl/config/samples/volume_bundle.yaml [hunk 9]
```diff
@@ -0,0 +1,16 @@
+---
+apiVersion: kjobctl.x-k8s.io/v1alpha1
+kind: VolumeBundle
+metadata:
+  name: sample-volume-bundle
+spec:
+  volumes:
+    - name: foo
+      hostPath:
+        path: /foo
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/config/samples/volume_bundle.yaml]: Done

**Review by trasc — COMMENTED** [review]
/lgtm

**Comment by mwysokin** [comment]
Just your friendly neighborhood shadow reviewer here. 

Overall, the code looks good, but I'm a bit concerned about some of the modules being used. They're a few years old and don't seem to be actively maintained anymore.  Not sure why we're going with these specific choices, but I'm wondering if there's any risk in using outdated dependencies for a brand new component.

Also, the examples seem to be copied directly from KubeRay and are referencing some pretty old Python packages (which can sometimes lead to issues with unsupported/outdated TLS certificates because ). I get that these examples might be pulled from another project, but I'm not entirely sure it's a good idea to just copy and paste them into our own without careful consideration.

One last thing: should we maybe mention somewhere that these examples are straight from KubeRay (https://raw.githubusercontent.com/ray-project/kuberay/v1.1.1/ray-operator/config/samples/ray-job.sample.yaml)? Just to be transparent.

**Comment by mbobrovskyi** [comment]
> Overall, the code looks good, but I'm a bit concerned about some of the modules being used. They're a few years old and don't seem to be actively maintained anymore. Not sure why we're going with these specific choices, but I'm wondering if there's any risk in using outdated dependencies for a brand new component.

Sorry, what dependencies do you mean exactly?

**Comment by mwysokin** [comment]
> > Overall, the code looks good, but I'm a bit concerned about some of the modules being used. They're a few years old and don't seem to be actively maintained anymore. Not sure why we're going with these specific choices, but I'm wondering if there's any risk in using outdated dependencies for a brand new component.
> 
> Sorry, what dependencies do you mean exactly?

The ones I made comments for so: `groupcache`, `beorn7/perks`, `matttproud/golang_protobuf_extensions`.

**Comment by mbobrovskyi** [comment]
> @mbobrovskyi The ones I made comments for so: groupcache, beorn7/perks, matttproud/golang_protobuf_extensions.

I think I got what you mean. There is a list of dependencies that are automatically added by `go mod tidy` and used by other dependencies. So it's okay that there're old and we don't need to change it to avoid conflicts.

File: cmd/experimental/kjobctl/config/samples/volume_bundle.yaml [hunk 10]
```diff
@@ -0,0 +1,54 @@
+---
+apiVersion: kjobctl.x-k8s.io/v1alpha1
```
Comment (mimowo) [inline cmd/experimental/kjobctl/config/samples/volume_bundle.yaml:2]: How is this file loaded? Does it require a manual non-documened step? If documented, please point me where is the command which applies this file.
Also, can we can name it "ray_volume_bundle.yaml` to reflect this is only for the Ray mode?

File: cmd/experimental/kjobctl/config/samples/application_profile.yaml [hunk 11]
```diff
@@ -45,3 +48,95 @@ template:
                 cpu: "1"
                 memory: "200Mi"
         restartPolicy: OnFailure
+---
+apiVersion: kjobctl.x-k8s.io/v1alpha1
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/config/samples/application_profile.yaml:53]: Done

File: cmd/experimental/kjobctl/config/samples/volume_bundle.yaml (hunk 10 above)
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/config/samples/volume_bundle.yaml]: > How is this file loaded?

It should be included on volumeBundles in ApplicationProfile. I added it on `application_profile.yaml`. Thanks.

Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/config/samples/volume_bundle.yaml]: > Also, can we can name it "ray_volume_bundle.yaml` to reflect this is only for the Ray mode?

Renamed.

Comment (mimowo) [reply cmd/experimental/kjobctl/config/samples/volume_bundle.yaml]: Thanks!

**Comment by mimowo** [comment]
> @mbobrovskyi The ones I made comments for so: `groupcache`, `beorn7/perks`, `matttproud/golang_protobuf_extensions`.

I think we are ok with these dependencies since they already are part of the main [go.mod](https://github.com/kubernetes-sigs/kueue/blob/main/go.mod).

**Comment by mimowo** [comment]
/lgtm
/approve
Thanks!

**Pull request merged**